source/opt/local_redundancy_elimination.h

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_OPT_LOCAL_REDUNDANCY_ELIMINATION_H_
+#define LIBSPIRV_OPT_LOCAL_REDUNDANCY_ELIMINATION_H_
+
+#include "ir_context.h"
+#include "pass.h"
+#include "value_number_table.h"
+
+namespace spvtools {
+namespace opt {
+
+// See optimizer.hpp for documentation.
+class LocalRedundancyEliminationPass : public Pass {
+ public:
+  const char* name() const override { return "local-redundancy-elimination"; }
+  Status Process(ir::IRContext*) override;
+
+ private:
+  // Deletes instructions in |block| whose value is in |vnTable| or is computed
+  // earlier in |block|.  The values computed in |block| are added to |vnTable|.
+  // |value_to_ids| is a map from value number to the result ids known to
+  // contain those values.  One value needs to map to multiple ids because the
+  // ids may contain the same value, but have different decorations.
```

Comment:
Perhaps do an actual map?  IDs can be sparse (perhaps it doesn't really matter, I'm not sure).

Reply:
The values are the value numbers in the value table.  Those numbers start at 1 and are incremented every time we need a new value number, they will generally be dense.  Right now the numbers are being reset for every basic block.  For the global version, it expect the numbers to be reset for every function, so they will still be fairly dense.